document/core/exec/instructions.rst

```diff
@@ -254,23 +254,474 @@ Reference Instructions
 
 1. Assert: due to :ref:`validation <valid-ref.is_null>`, a :ref:`reference value <syntax-ref>` is on the top of the stack.
 
-2. Pop the value :math:`\val` from the stack.
+2. Pop the value :math:`\reff` from the stack.
 
-3. If :math:`\val` is :math:`\REFNULL~\X{ht}`, then:
+3. If :math:`\reff` is :math:`\REFNULL~\X{ht}`, then:
 
    a. Trap.
 
-4. Push the value :math:`\val` back to the stack.
+4. Push the value :math:`\reff` back to the stack.
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   \reff~\REFASNONNULL &\stepto& \TRAP
+     & (\iff \reff = \REFNULL~\X{ht}) \\
+   \reff~\REFASNONNULL &\stepto& \reff
+     & (\otherwise) \\
+   \end{array}
+
+
+.. _exec-ref.eq:
+
+:math:`\REFEQ`
+..............
+
+1. Assert: due to :ref:`validation <valid-ref.eq>`, two :ref:`reference values <syntax-ref>` are on the top of the stack.
+
+2. Pop the value :math:`\reff_2` from the stack.
+
+3. Pop the value :math:`\reff_1` from the stack.
+
+4. If :math:`\reff_1` is the same as :math:`\reff_2`, then:
+
+   a. Push the value :math:`\I32.\CONST~1` to the stack.
+
+5. Else:
+
+   a. Push the value :math:`\I32.\CONST~0` to the stack.
+
+.. math::
+   \begin{array}{lcl@{\qquad}l}
+   \reff_1~\reff_2~\REFEQ &\stepto& (\I32.\CONST~1)
+     & (\iff \reff_1 = \reff_2) \\
+   \reff_1~\reff_2~\REFEQ &\stepto& (\I32.\CONST~0)
+     & (\iff \reff_1 \neq \reff_2) \\
+   \end{array}
+
+
+.. _exec-ref.test:
+
+:math:`\REFTEST~\X{rt}`
+.......................
+
+1. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+
+2. Let :math:`\X{rt}_1` be the :ref:`reference type <syntax-reftype>` :math:`\insttype_{F.\AMODULE}(\X{rt})`.
+
+3. Assert: due to :ref:`validation <valid-ref.test>`, :math:`\X{rt}_1` is :ref:`closed <type-closed>`.
```

Comment:
It's an (indirect) consequence of validation that `clos_{moduleinst}(rt)` is indeed closed, because validation of the instruction also validates `rt`, and that in turn validates that it only contains type indices bound in the module.